### PR TITLE
New AWS IoT Core client based on work from Nantis GmbH

### PR DESCRIPTION
**Pull request by LDong-Arm**
Bora Özgen (@boraozgen) from Nantis GmbH has kindly created https://github.com/Nantis-GmbH/mbed-aws-client which is based on coreMQTT, coreJSON and Device-Shadow-for-AWS-IoT-embedded-sdk from AWS and FreeRTOS. This supports the publishing and receiving of MQTT messages, as well as the Device Shadow service.

This PR imports that into the official repository maintained by Mbed OS, and makes some adjustments to align with Mbed OS requirements: coding standard, CMake refinement, etc. See the commit history and messages for what we have changed.

The previous Mbed OS port for the AWS SDK (up to 5f422c94) is based on [a deprecated version of the AWS SDK](https://github.com/aws/aws-iot-device-sdk-embedded-C/tree/v4_beta_deprecated) which has a completely different code base. This PR removes the old port, but existing users can continue to use that SHA-1, while migration to the new client is highly recommended.

### CODE REVIEW

**Review by evedon — COMMENTED** [review]
This looks very good. I only have a few minor comments.
It would be great to see error handling implemented in a future revision.

File: source/AWSClient.cpp [hunk 1]
```diff
@@ -0,0 +1,728 @@
...
+}
+
+void AWSClient::eventCallbackStatic(MQTTContext_t *pMqttContext,
+                                    MQTTPacketInfo_t *pPacketInfo,
+                                    MQTTDeserializedInfo_t *pDeserializedInfo)
+{
+    // TODO error handling
+
+    MBED_ASSERT(pMqttContext != NULL);
+    MBED_ASSERT(pPacketInfo != NULL);
```
[the remark this answers is not in the document]
Comment (LDong-Arm) [reply source/AWSClient.cpp:93]: Done

File: source/AWSClient.cpp [hunk 2]
```diff
@@ -0,0 +1,728 @@
...
+                    awsClient.shadowUpdateAccepted = true;
+                    break;
+
+                case ShadowMessageTypeUpdateRejected:
+                    tr_warn("/update/rejected json payload: %.*s", pPublishInfo->payloadLength, (const char *)pPublishInfo->pPayload);
+                    awsClient.shadowUpdateAccepted = false;
+                    break;
+
+                default:
+                    tr_warn("Received unexpected shadow message");
```
[the remark this answers is not in the document]
Comment (LDong-Arm) [reply source/AWSClient.cpp]: Done

File: source/AWSClient.cpp [hunk 3]
```diff
@@ -0,0 +1,728 @@
...
+    if ((pPacketInfo->type & 0xF0U) == MQTT_PACKET_TYPE_PUBLISH) {
+        MBED_ASSERT(pDeserializedInfo->pPublishInfo != NULL);
+        MQTTPublishInfo_t *pPublishInfo = pDeserializedInfo->pPublishInfo;
+
+        /* Let the Device Shadow library tell us whether this is a device shadow message. */
+        ShadowMessageType_t messageType = ShadowMessageTypeMaxNum;
+        const char *pThingName = NULL;
+        uint16_t thingNameLength = 0U;
+
+#if MBED_CONF_AWS_CLIENT_SHADOW
```
[the remark this answers is not in the document]
Comment (LDong-Arm) [reply source/AWSClient.cpp]: Done

**Review by evedon — APPROVED** [review]

**Pull request merged**